spring-cloud-dataflow-aggregate-task/src/main/java/org/springframework/cloud/dataflow/aggregate/task/impl/AggregateDataFlowTaskExecutionQueryDao.java

```diff
@@ -143,18 +144,54 @@ public class AggregateDataFlowTaskExecutionQueryDao implements DataflowTaskExecu
 			+ "PARENT_EXECUTION_ID, SCHEMA_TARGET"
 			+ " from AGGREGATE_TASK_EXECUTION";
 
+	private static final String GET_EXECUTION_BY_NAME_COMPLETED_BEFORE_END_TIME = "SELECT TASK_EXECUTION_ID,"
+			+ "START_TIME, END_TIME, TASK_NAME, EXIT_CODE,"
+			+ "EXIT_MESSAGE, ERROR_MESSAGE, LAST_UPDATED, EXTERNAL_EXECUTION_ID,"
+			+ "PARENT_EXECUTION_ID, SCHEMA_TARGET"
+			+ " from AGGREGATE_TASK_EXECUTION where TASK_NAME = :taskName AND END_TIME IS NOT NULL AND END_TIME < :endTime";
+
+	private static final String GET_EXECUTIONS_COMPLETED_BEFORE_END_TIME = "SELECT TASK_EXECUTION_ID,"
+			+ "START_TIME, END_TIME, TASK_NAME, EXIT_CODE,"
+			+ "EXIT_MESSAGE, ERROR_MESSAGE, LAST_UPDATED, EXTERNAL_EXECUTION_ID,"
+			+ "PARENT_EXECUTION_ID, SCHEMA_TARGET"
+			+ " from AGGREGATE_TASK_EXECUTION where END_TIME IS NOT NULL AND END_TIME < :endTime";
+
+	private static final String GET_EXECUTION_BY_NAME_BEFORE_END_TIME = "SELECT TASK_EXECUTION_ID,"
+			+ "START_TIME, END_TIME, TASK_NAME, EXIT_CODE,"
+			+ "EXIT_MESSAGE, ERROR_MESSAGE, LAST_UPDATED, EXTERNAL_EXECUTION_ID,"
+			+ "PARENT_EXECUTION_ID, SCHEMA_TARGET"
+			+ " from AGGREGATE_TASK_EXECUTION where TASK_NAME = :taskName AND END_TIME < :endTime";
+
+	private static final String GET_EXECUTIONS_BEFORE_END_TIME = "SELECT TASK_EXECUTION_ID,"
+			+ "START_TIME, END_TIME, TASK_NAME, EXIT_CODE,"
+			+ "EXIT_MESSAGE, ERROR_MESSAGE, LAST_UPDATED, EXTERNAL_EXECUTION_ID,"
+			+ "PARENT_EXECUTION_ID, SCHEMA_TARGET"
+			+ " from AGGREGATE_TASK_EXECUTION where END_TIME < :endTime";
+
 	private static final String TASK_EXECUTION_COUNT = "SELECT COUNT(*) FROM "
 			+ "AGGREGATE_TASK_EXECUTION ";
 
+	private static final String TASK_EXECUTION_COUNT_BEFORE_END_TIME = "SELECT COUNT(*) FROM "
+			+ "AGGREGATE_TASK_EXECUTION WHERE END_TIME < :endTime";
+
 	private static final String TASK_EXECUTION_COUNT_BY_NAME = "SELECT COUNT(*) FROM "
 			+ "AGGREGATE_TASK_EXECUTION where TASK_NAME = :taskName";
 
+	private static final String TASK_EXECUTION_COUNT_BY_NAME_AND_BEFORE_END_TIME = "SELECT COUNT(*) FROM "
+			+ "AGGREGATE_TASK_EXECUTION where TASK_NAME = :taskName AND END_TIME < :endTime";
```

Comment:
What happens when end time is null? If it is included in the result then we could cleanup still running tasks. If it is excluded then it does not make any sense to pass in a `onlyCompleted` flag.

Reply:
The end time should never be null and is passed in from the calling methods which calculate the endTime. But you are right. I added `@NotNull` to the methods to ensure that the end time is always given.

Reply:
Sorry, I meant the `END_TIME` on the record which would make `END_TIME < :endTime` (ie. null < '2023-08-30 23:59:59').

Reply:
I fixed it accordingly as I already saw that you could only meant `END_TIME` and not `onlyCompleted` as this was not present in the query 👍